File: java/Hot100/Matrix/SetMatrixZeroes.java
package Hot100.Matrix;

import java.util.HashSet;
import java.util.Set;

/**
 * Given an m x n integer matrix matrix, if an element is 0, set its entire row and column to 0's.
 *
 * You must do it in place.
 */
public class SetMatrixZeroes {
    public void setZeroes(int[][] matrix) {
        int row= matrix.length;
        int col=matrix[0].length;
        Set<Integer> rowSet=new HashSet<>();
        Set<Integer> colSet=new HashSet<>();

        //record zeroes first, not to cover original zeroes
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                int num=matrix[i][j];
                if(num==0){
                    rowSet.add(i);
                    colSet.add(j);
                }
            }
        }

        for (Integer i:rowSet) {
            for (int j = 0; j < col; j++) {
                matrix[i][j]=0;
            }
        }

        for (Integer i:colSet) {
            for (int j = 0; j < row; j++) {
                matrix[j][i]=0;
            }
        }
    }
}

File: python/matrix/set_matrix_zeroes.py
# Given an m x n integer matrix matrix, if an element is 0, set its entire row and column to 0's.
# You must do it in place.

class Solution:
    def setZeroes(self, matrix: list[list[int]]) -> None:
        m = len(matrix)
        n = len(matrix[0])
        first_row_has_zero = any(matrix[0][j] == 0 for j in range(n))
        first_col_has_zero = any(matrix[i][0] == 0 for i in range(m))

        # Use the first row and column to mark zeros
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0

        # Set zeros based on marks in the first row and column
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

        # Set the first row to zero if needed
        if first_row_has_zero:
            for j in range(n):
                matrix[0][j] = 0

        # Set the first column to zero if needed
        if first_col_has_zero:
            for i in range(m):
                matrix[i][0] = 0